### ISSUE TITLE
Add ability to deploy custom elastic-agents

### ISSUE THREAD

**Issue by marc-gr** [elastic/elastic-package#787]
There are some integrations that might require elastic-agents with custom configurations for them or their container ie: winlogbeat requires a windows container, auditbeat requires special container capabilities, etc.

I initially created https://github.com/elastic/elastic-package/pull/786 that adds ability to deploy custom agents as test services, there is still missing code specific to deal with the windows scenario.

I open this thread to discuss other approaches that might avoid adding the complexity to the test runner if possible.

**Comment by mtojek** [comment]
pinging @jsoriano for his thoughts around this and https://github.com/elastic/elastic-package/pull/786.

**Comment by jsoriano** [comment]
Another option could be to move agent initialization to the system test runner, and remove it from the default stack definition. Having it in the runner would allow it to have full control of the started agents, allowing to start them with different options, and handling platform-specific needings, as could be the case of Windows.
It could also allow to start different tests with different agent configuration, for example to test different auditbeat configurations with different capabilities.

We already have a custom agent for the Kubernetes service deployer, we could follow a similar strategy on any other deployer.
There could be general options, such as options to select the version to use, or things like capabilities to add. And there could be platform specific options, that would also allow things like selecting daemonset vs deployment in Kubernetes (https://github.com/elastic/elastic-package/issues/465). On a second iteration these options could be easily overridden with flags or environment variables.

If we remove agent from the default stack definition, we could still have a stack subcommand to start agents for manual tests. Something like this could also cover https://github.com/elastic/elastic-package/issues/548.

I think this could be a more future-proof option, but it can require an important effort.

---

And another option for the use case of starting an agent but no service, can be to add a [new "system" deployer](https://github.com/elastic/elastic-package/blob/bbdb1f3238ac78b17fdeedc72a75c2e4fef52997/internal/testrunner/runners/system/servicedeployer/factory.go#L40), that just starts an agent with a given configuration, intended for system-level monitoring. This could help with packages for auditbeat or for the system module itself. This could be extended in the future to start completely different OSs using VMs.

This would be more in line of https://github.com/elastic/elastic-package/pull/786, but without needing to hack over the current test runner and compose deployer.

**Comment by mtojek** [comment]
> Another option could be to move agent initialization to the system test runner, and remove it from the default stack definition. Having it in the runner would allow it to have full control of the started agents, allowing to start them with different options, and handling platform-specific needings, as could be the case of Windows.
It could also allow to start different tests with different agent configuration, for example to test different auditbeat configurations with different capabilities.

There are two constraints related to this approach:
1. Don't forget that this is the mode we also use for development purposes. You can simply start the stack and have everything ready. It is really convenient. 
2. Agent enrollment with fleet server takes time. We considered this option at the early stage and decided to follow the "enroll once" approach at startup. It's also easier to debug if you have the agent instance present, not wiped out.

> We already have a custom agent for the Kubernetes service deployer, we could follow a similar strategy on any other deployer.
There could be general options, such as options to select the version to use, or things like capabilities to add. And there could be platform specific options, that would also allow things like selecting daemonset vs deployment in Kubernetes (https://github.com/elastic/elastic-package/issues/465). On a second iteration these options could be easily overridden with flags or environment variables.

I like the approach of having the custom agent setup. You're right that we could apply similar logic as for `kind`, to spawn a new agent. This way we don't need to modify test runners at all. Most likely we may need two setups: `custom image properties` and `windows`.

> If we remove agent from the default stack definition, we could still have a stack subcommand to start agents for manual tests. Something like this could also cover https://github.com/elastic/elastic-package/issues/548.

I had that in mind before, hence the issue, but always considered its complexity as +Inf. Maybe we can evaluate it as a good first issue and "rebuild the stack command"? 

---

To sum up, my vote would go to custom agent setup.

**Comment by jsoriano** [comment]
> > We already have a custom agent for the Kubernetes service deployer, we could follow a similar strategy on any other deployer.
> > There could be general options, such as options to select the version to use, or things like capabilities to add. And there could be platform specific options, that would also allow things like selecting daemonset vs deployment in Kubernetes (#465). On a second iteration these options could be easily overridden with flags or environment variables.
> 
> I like the approach of having the custom agent setup. You're right that we could apply similar logic as for `kind`, to spawn a new agent. This way we don't need to modify test runners at all. Most likely we may need two setups: `custom image properties` and `windows`.

Could this be done without modifying runners?

**Comment by mtojek** [comment]
Yes, I think so. Same way as closed most of the changes for Kubernetes service deployer in this [file](https://github.com/elastic/elastic-package/blob/main/internal/testrunner/runners/system/servicedeployer/kubernetes.go). There might be one inconvenience, the agent will be deployed during the first run of the system test.

**Comment by jsoriano** [comment]
Ah ok, but it would be modifying service deployers. Would you prefer to add an agent to the compose deployer, or to add a new deployer for these use cases?

**Comment by mtojek** [comment]
> Would you prefer to add an agent to the compose deployer

It looks like it depends on the final infrastructure setup. Not sure if that option will work for @marc-gr and Windows containers.

> add a new deployer for these use cases?

This option seems to be pluggable and flexible in terms of specific configuration properties or OS-specific logic. It has also an extra benefit, it will prevent copying a custom agent code to multiple places.

I'm thinking now if we aren't close to introducing a feature of using an agent under development. This way you could use even standalone builds. Maybe we should implement a proxy instead :)

**Comment by jsoriano** [comment]
Discussed offline about this with Marc, he is going to explore the option of implementing something like https://github.com/elastic/elastic-package/pull/786, but as a new deployer, so the runner is not modified. This could cover the current auditbeat needings.

We also discussed that probably we need something like vms for system tests, this will be neccesary to support running tests with windows, or even with linux if not enough privileges can be granted with containers for some use cases.
We could run these tests on specialized CI workers, as in https://github.com/elastic/integrations/pull/1713 for Windows, but it'd be nice to have something in `elastic-package` so developers working on Mac/Linux can also run these tests locally.

**Comment by cmacknz** [comment]
> but it'd be nice to have something in elastic-package so developers working on Mac/Linux can also run these tests locally.

For Linux https://multipass.run/ is a good cross-platform solution as long as you are fine with only supporting Ubuntu VMs. For Windows there is no cross-platform equivalent, you have to provision cloud VMs.

This is generally what we do in the Elastic Agent test framework, https://github.com/elastic/elastic-agent/blob/main/docs/test-framework-dev-guide.md. You can test locally against multipass Ubuntu VMs, otherwise we are provisioning Linux and Windows machines in the cloud. MacOS VM support is TBD.

It would be good if we could align the provisioning here with the agent framework so we aren't maintaining this functionality twice. The only quirk with the agent test framework is it uses https://github.com/adam-stokes/ogc for provisioning, we'd prefer to use Terraform but we haven't gotten that implemented yet. https://github.com/elastic/elastic-agent/issues/2935

**Comment by cmacknz** [comment]
CC @blakerouse

**Comment by mrodm** [comment]
With the latest Pull Requests merged into `elastic-package` linked to this issue, there has been added support to run independent Elastic Agents mainly to run system tests. Currently, this is enabled when an environment variable `ELASTIC_PACKAGE_TEST_ENABLE_INDEPENDENT_AGENT` exists and has the value `true` (as a feature flag).  Example:
```shell
elastic-package stack up -v -d
export ELASTIC_PACKAGE_TEST_ENABLE_INDEPENDENT_AGENT=true
elastic-package test system -v
``` 

If enabled this feature, then that means that for each data stream and each configuration file defined in the packages, a new Elastic Agent is created/enrolled and destroyed when that test finishes.

In this PR #9660, this feature has been enabled to test all packages enrolling a new Elastic Agents for each test. Thereby, Elastic Agent created during `elastic-package stack up` command is not used at all.

One side-effects is that packages with the largest number of datasets (and configuration files) in the integrations repository last longer to finish all tests:

| Package | Running time using Elastic Agent stack | Running time using Independent Elastic Agent |
| --- | --- | --- |
| aws | 1h 20min | 1h 40min |
| network_traffic | 1h 20min | 2h 20min |
| zeek | 1h 10min | 2h 40 min |

To allow testing all packages successfully, it was also needed to upgrade docker daemon to 26.1.0 version #9710. There was a package (`oracle_weblogic`) with some network issues between containers (e.g. requests from Elastic Agent to the service container timeouts) when independent Elastic Agents were enabled and docker version was 24.0.7 (example [build link](https://buildkite.com/elastic/integrations/builds/10761)). Those issues were solved if docker daemon was updated.

As part of this process, if packages have set `agent.privileges.root` as true in their package manifest, then the container created for the Elastic Agent adds `user: root` to run the main process (#1789 related to the issue #1586).

What should be the next steps now? Should we enable just for some packages in the integration repository this new process of running Elastic Agents ? @elastic/ecosystem

**Comment by jsoriano** [comment]
> One side-effects is that packages with the largest number of datasets (and configuration files) in the integrations repository last longer to finish all tests

I wouldn't worry too much about this by now because current time is already too much and we should do something about this in any case. The new approach goes in the direction of allowing parallelization of tests at some point, so it is a good step forward even if it looks so worse now. Said that, it would be better if we don't enable this for these packages by now.

> As part of this process, if packages have set `agent.privileges.root` as true in their package manifest, then the container created for the Elastic Agent adds `user: root` to run the main process (#1789 related to the issue #1586).

> What should be the next steps now? Should we enable just for some packages in the integration repository this new process of running Elastic Agents ?

Would it be an option to enable independent agents only for packages using `agent.privileges.root` by now? This would allow to close #1586, and we would be testing independent agents in the wild with some packages.

**Comment by mrodm** [comment]
> Would it be an option to enable independent agents only for packages using `agent.privileges.root` by now? This would allow to close #1586, and we would be testing independent agents in the wild with some packages.

Working on this PR https://github.com/elastic/elastic-package/pull/1815 to enable independent Elastic Agents directly in elastic-package. If needed, that could be disabled by setting as false the env. var ELASTIC_PACKAGE_TEST_ENABLE_INDEPENDENT_AGENT.

**Comment by mrodm** [comment]
Added support in `elastic-package` to define custom script into the boot (and start) process:
https://github.com/elastic/elastic-package/pull/1822

This is needed at least for the `oracle` package that uses custom agents with a custom Dockerfile for the Elastic Agent.

**Comment by mrodm** [comment]
With the latest PRs merged, it looks like that there is support to run all our current use cases:
- Elastic Agents are now started/enrolled for each test, and when the test finishes that Elastic Agent is unenrolled permanently.
- Settings that can be customized in configuration files:
    - Set another user to run Elastic Agent.
    - Enable sharing the PID address space between container and the host operating system.
    - Expose ports to access Elastic Agent.
    - Add Linux capabilities
        - All capabilities are removed by default.
    - Provide option to run custom scripts to prepare the environment as it is required for each package where runs Elastic Agent.
- Add support to run system tests with Elastic Serverless projects.

Do you think of any other setting or use case that needs to be take into consideration here? @jsoriano 

Along with all these changes, a new PR has been created to update docs about system tests too: #1830 

What should be the next steps from here? 
- Keep using Elastic Agent from stack (and previous custom agents) and just use independent Elastic Agents for packages that require root privileges and close this issue as it is.
- Enable independent Elastic Agents by default in `elastic-package`?
    - Require to update packages using custom agents
    - Require to change `elastic-package` default behavior
    - Packages like `aws` or `zeek` would last 2h 30-40 min approx
- Enable independent Elastic Agents in integrations repository using just the environment variable (no change in `elastic-package`)
    - Require to update packages using custom agents
    - Packages like `aws` or `zeek` would last 2h 30-40 min approx

I created this PR with changes required to use new Elastic Agents all packages in the integrations repository: https://github.com/elastic/integrations/pull/9862

**Comment by jsoriano** [comment]
> Do you think of any other setting or use case that needs to be take into consideration here? @jsoriano

We can add them as needed, so far I don't think anything else has been needed or requested.

> Enable independent Elastic Agents by default in elastic-package? 

Yes, we have to go in this direction, the main blocker I see now is the slower time, we have to find strategies to mitigate this.

**Comment by jlind23** [comment]
> Yes, we have to go in this direction, the main blocker I see now is the slower time, we have to find strategies to mitigate this.

I agree with @jsoriano on this, this will help engineers to test many more integrations. For the slowness, I don't think it is a problem in CI but it might be a drawback while testing locally, do we already have a lead on what could be improved here?
Anyway once the independent Elastic Agent becomes the default I think we can close this issue as done and create follow up tasks if and when required.

**Comment by jsoriano** [comment]
> For the slowness, I don't think it is a problem in CI but it might be a drawback while testing locally

I kind of think the opposite, in CI everything is run, and for some packages this is going from about 1h30m to 2h40m. Locally one can run only the tests for one data stream, and even have the scenario pre-started, and have a much quicker feedback.

In CI we were already considering the current times a problem for some packages, one hour more can be too much.

> do we already have a lead on what could be improved here?

Now that things are more isolated with the independent agents, I think that we can start parallelizing things that we do sequentially now:
* Parallelize tests per data stream. We have to be careful with the peak resources, but having only a factor of two could already lead to similar or better times than the ones we have now.
* Add a background process that keeps a pool of a pre-started agents when testing packages with many data streams.
* Parallelize parts of the test itself. For example we could likely install the package and create the policy while we are starting the agent.

> Anyway once the independent Elastic Agent becomes the default I think we can close this issue as done and create follow up tasks if and when required.

+1

**Comment by jsoriano** [comment]
Well, for slowness, another middle-ground temporary step could be to enable independent agents for all packages, except for the ones with more than N data streams or tests. This could be probably done soon.

**Comment by mrodm** [comment]
Setting as default running independent Elastic Agents, should be just change this line
https://github.com/elastic/elastic-package/blob/7b156a51b8db9f300d4ba2fb67d6bcea9b744b42/cmd/testrunner.go#L192-L194

to be always true, and keep the environment variable in case it is required to be disabled for some reason.

Looking into options I think also in trying to run tests in parallel. There are some changes required for that, from the top of my head
- Revisit compose project names to not use the same project name in different runs
    - A parallel build/run could tear down the service since it is the same project name ?
    - It could be added the RunID that it is generated for each test

**Comment by jlind23** [comment]
Parallelisation looks like the way to go anyway and sorry for my comment I didn't know that the CI time already was a burden for integrations developers. 🤐

**Comment by mrodm** [comment]
> Well, for slowness, another middle-ground temporary step could be to enable independent agents for all packages, except for the ones with more than N data streams or tests. This could be probably done soon.

This could be an option as a trade-off to start using these new Elastic Agents.

However, not sure if this logic would be better to add into into `elastic-package` CLI directly or it should be part of the CI scripts in the integrations repository (to define `ELASTIC_PACKAGE_TEST_ENABLE_INDEPENDENT_AGENT=false`).

There is one special consideration if we apply this. One of the packages with quite a few data streams is `network_traffic`. This package requires root privileges , so **it currently already runs independent Elastic Agents** to ensure it runs the Elastic Agent process as root.

This package contains 16 data streams running system tests and taking into account all the configuration files defined, so there are 91 runs. This is the package with more tests.

Latest executions of this package are between 2h 45min and 3h 
- https://buildkite.com/elastic/integrations/builds/11451#018f79e4-52c2-4db4-9f0f-7470a2a5a0c2
- https://buildkite.com/elastic/integrations/builds/11440#018f7817-b0f3-4247-9674-51014dd8fb62

These are the top 10 packages with more tests (ordered by number of configuration files defined):

| Package | Number system folders | Number configuration files | 
| --- | --- | --- |
| network_traffic | 16 | 91 |
| zeek | 43 | 81 |
| kubernetes | 25 | 25 |
| cloudflare_logpush | 18 | 18 |
| aws | 17 | 18 |
| google_workspace | 14 | 14 |
| elasticsearch | 13 | 14 |
| tanium | 6 | 12 |
| forgerock | 11 | 11 |
| darktrace | 3 | 11 |

**Comment by mrodm** [comment]
Created two PRs to check between the two options:
- https://github.com/elastic/integrations/pull/9885
    - It runs tests using independent Elastic Agents up until 50 tests
        - Packages with more than 50 packages use current Elastic Agent from stack.
    - If package requires root privileges, `elastic-package` still uses independent Elastic Agents for it:
        - `network_traffic` is one of them and it lasts between 2h 45min - 3h
- https://github.com/elastic/integrations/pull/9885
    - First steps to achieve parallel execution

**Comment by mrodm** [comment]
It has been observed that there is some flakiness when independent Elastic Agents are enabled. For instance `cloudflare` was failing randomly with `Could not find hits` error.

After some investigation, it has been checked that setting for the Buildkite steps Ubuntu 22.04 as image mitigated (solved?) these issues. 

Here it has been tested enabling independent Elastic Agents that packages like `cloudflare` were not failing in the latest builds: https://github.com/elastic/elastic-package/issues/787

Becuase of that, it has been created PR to set Ubuntu 22.04 as image in the Buildkite steps in charge of testing packages: https://github.com/elastic/integrations/pull/9912

**Comment by mrodm** [comment]
When enabling independent Elastic Agents, there are some packages that last around 3 hours to finish their tests (mainly system tests).

Added a new PR to allow creating a new Agent Policy per each test executed: https://github.com/elastic/elastic-package/pull/1866

This will allow us to:
- One step closer to be able to run in parallel system tests since every tests is going to use a different data stream to ingest docs.
- Reduce complexity when using stages in system tests (e.g. `--no-provision` flag)

**Comment by mrodm** [comment]
Two new PRs created to change how test runners work in `elastic-package`:
- #1895 
- #1898

These two PRs introduce two different interfaces to manage runners and tests:
- `Tester` interface:
    - it handles the execution of just one test with its own lifecycle
        - Each testrunner can define its own specific tests, for instance:
            - asset tests: just one test for all the package.
            - system tests: one test per configuration file and variant.
            - policy tests: one test per configuration file.
            - ...
- `TestRunner` interface:
    - it handles the creation (and destruction) of global resources required for tests.
    - it handles the creation of Tester instances, one per test defined in the package.

**Comment by mrodm** [comment]
Next step is adding support to run system tests in parallel in `elastic-package`.

This work is being done in two different PRs:
- https://github.com/elastic/package-spec/pull/759: allow to define a new global configuration file for tests.
- https://github.com/elastic/elastic-package/pull/1909: add support in elastic-package to run in parallel system tests

This will allow us to set system tests in parallel in packages with large number of system tests like `network_traffic` or `zeek`.

**Comment by mrodm** [comment]
Running some tests in this PR from integrations with just 2 packages (`network_traffic` and `zeek`) https://github.com/elastic/integrations/pull/10161

Comparing times among the different settings:

| Package | Sequential (stack Elastic Agent) | Sequential (independent Elastic Agents) | Up to 3 | Up to 5 | Up to 8 | 
| ---- | ---- | ---- | ---- | ---- | ---- | 
| network_traffic | 1h 20min | 2h 20min | 1h 7min | 42min | Error (timeouts) |
| zeek            | 1h 10min | 2h 40min | 1h 8min | 42min | Error (timeouts) |

CI builds: 
- Sequential times from this comment: https://github.com/elastic/elastic-package/issues/787#issuecomment-2090971992
- Up to 3 routines: https://buildkite.com/elastic/integrations/builds/12562
- Up to 5 routines: https://buildkite.com/elastic/integrations/builds/12568
- Up to 5 routines (2nd attempt): https://buildkite.com/elastic/integrations/builds/12575
- Up to 8 routines: https://buildkite.com/elastic/integrations/builds/12573

**Comment by mrodm** [comment]
I was wondering to close this issue once this PR (https://github.com/elastic/elastic-package/pull/1909) is merged.

All the support related to independent Elastic Agents and running system tests in parallel would be completed at that point.

It would be missing:
- release a new version of elastic-package and being integrated to integrations repository.
- enable independent Elastic Agents (through environment variable) and enable parallel system tests in `network_traffic` and `zeek` packages.

For that, it could be created a follow-up issue to enable those features in the integrations repository. There will be some packages to update while doing so. At least, `auditd_manager` and `oracle` (see related PoC PR about the changes required https://github.com/elastic/integrations/pull/9862). 

It could be created another issue to run the system tests using the independent Elastic Agents by default. Could this be done as part of a different issue too?

However, that means that developers would be triggering the tests using the Elastic Agent from the stack but the CI would be using the new independent Elastic Agents. If they want to be running independent Elastic Agents should be setting the environment variable: `ELASTIC_PACKAGE_TEST_ENABLE_INDEPENDENT_AGENT=true`


WDYT about closing this one (once [the PR is merged](https://github.com/elastic/elastic-package/pull/1909)) in favor of creating those new issues? @jsoriano @kpollich

**Comment by mrodm** [comment]
Just to add to the previous comment, it should be updated the docs too about these new settings.

I'll update the current PR with the changes required about the docs:
https://github.com/elastic/elastic-package/blob/main/docs/howto/system_testing.md#running-system-tests-with-independent-elastic-agents-in-each-test-technical-preview

**Comment by jsoriano** [comment]
> I was wondering to close this issue once this PR (#1909) is merged.

Yep, I mostly agree with closing this once we can enable independent agents more generally. But please take into account that one the original motivations for this issue was to be able to run winlog tests, and we are still unable to run Windows agents for this.
If we close this issue, please ensure that we keep some issue open for use cases on different operating systems.

**Comment by mrodm** [comment]
> But please take into account that one the original motivations for this issue was to be able to run winlog tests, and we are still unable to run Windows agents for this.
> If we close this issue, please ensure that we keep some issue open for use cases on different operating systems.

That's true, I could keep this issue open (since there are other issues already linked to this one) even if the above mentioned PRs are merged, until we could find time to work on adding support to run Elastic Agents in other OS or runtimes.

**Comment by mrodm** [comment]
Created package-spec release 3.2.0 (https://github.com/elastic/package-spec/releases/tag/v3.2.0) that includes the definition of the new configuration files to enable or not system parallel tests.

**Comment by mrodm** [comment]
Once it is merged the support to run the test using independent Elastic Agents as well as to run system tests in parallel in `elastic-package`, I created this follow-up issue to enable these features in the integrations repository:
https://github.com/elastic/integrations/issues/10201

**Comment by kpollich** [comment]
Thanks for providing a summary of where we are today, @mrodm. I'm moving this into a quality sprint for now as we'll need to dedicate a large amount of time here if we prioritize adding cross-platform support to this new type of test.

**Comment by mrodm** [comment]
A use case that this feature could be helpful would be for the `system_audit` package (running Elastic agent in different VMs).

This would allow to run the system tests with agents running in different Linux OS other than Ubuntu, e.g. Fedora. So it could be tested that it can collect the required logs from rpm package manager.
Related to https://github.com/elastic/integrations/issues/11000
